types/underscore/index.d.ts

```diff
@@ -115,6 +117,38 @@ declare module _ {
         : V extends Dictionary<infer T> ? T
         : never;
 
+    // given a union type like { a: string } | { b: number } | undefined, generates a union of all the keys of all the types in the union
+    type KeysOfUnion<T> = T extends any ? keyof T : never;
+
+    // given a union type like { a: string } | { a: number } | { b: boolean } and a property name,
+    // generates a union of all the types implemented by that property name and adds undefined
+    // if any type in the union doesn't implement the property
+    // e.g. TypesOfUnionProperty<{ a: string } | { a: number } | { b: boolean }, 'a'> => string | number | undefined
+    type TypesOfUnionProperty<T, K extends KeysOfUnion<T>> =
+        // iterate over each type T in a type union
+        T extends any
+        // check whether the property is implemented by T, using Extract to get around K possibly not being in keyof T
+        ? T[Extract<keyof T, K>] extends never
+        // if T does not implement property K, add undefined to the resulting type union
+        ? undefined
+        // if T does implement property K, add type T[K] to the resulting type union
+        : T[Extract<keyof T, K>]
+        // this should never actually be evaluated since all types extend any
+        : never;
+
+    // '& object' prevents strings from being matched by list checks
+    type ShallowFlattenedList<T> = T extends List<infer TItem> & object ? TItem : T;
```

Comment:
It would probably be better to call this `ShallowFlattenedListItem<T>` since what I'm really getting back here is the item type for the list.

Reply:
Per [this conversation](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r448051126) I went with `ListItemOrType` here.